libsolidity/analysis/TypeChecker.cpp

```diff
@@ -355,8 +355,16 @@ bool TypeChecker::visit(FunctionDefinition const& _function)
 		{
 			if (!type(var)->canLiveOutsideStorage() && _function.isPublic())
 				m_errorReporter.typeError(var.location(), "Type is required to live outside storage.");
-			if (_function.isPublic() && !(type(var)->interfaceType(isLibraryFunction).get()))
-				m_errorReporter.fatalTypeError(var.location(), "Internal or recursive type is not allowed for public or external functions.");
+			if (_function.isPublic())
+			{
+				auto iType = type(var)->interfaceType(isLibraryFunction);
+
+				if (!iType.get())
+				{
+					auto msg = iType.message().empty() ? "Invalid type! ()" : iType.message();
```

Comment:
Where does `"Invalid type! ()"` appear? Shouldn't it be an assertion? If not, what is the idea behind the `()`?

Reply:
I forgot to take that out, it was merely to differentiate this from another (equally phrased) error. We can probably turn it into an assert now, yes.